src/mlpack/tests/matrix_completion_test.cpp

```diff
@@ -64,4 +66,54 @@ BOOST_AUTO_TEST_CASE(UniformMatrixCompletionSDP)
   }
 }
 
+
+/**
+ * Matrix completion test with Frank-Wolfe Type solver.
+ *
+ * The test data is the same as the previous test.
+ *
+ * The matrix is recovered by solving the model:
+ *
+ * min \sum_{i,j} (X_ij - M_ij)^2
+ * s.t. ||X||_* <= tau
+ *
+ * where ||X||_* is the nuclear norm of X.
+ *
+ * The convergence is really slow.
```

Comment:
I see that in the test times, like your comment says, the test takes significantly longer than the SDP solver.  Do we have a way to reduce the runtime of this test?  Maybe the tolerance can be reduced?  The second question is, do we know situations where the FW solver will outperform the existing SDP solver?

Reply:
It is known that the convergence rate of FW is only O(1/k). So we will need much more iterations to get a small error. I guess there is no way to fix it. The benefit of FW is that it is really fast in each iteration. This could be seen for large scale problems (when the matrix size is huge). In that case, each iteration of SDP is very slow. So the comparison between FW and SDP is:
(fast in each step) x (more iterations)     versus    (slow in each step) x (less iterations)

In the unit test, the problem size is really small, so SDP is much faster to beat FW. I guess if you want to see the situation that FW outperforms SDP, we need to test with large scale problems, where SDP becomes infeasible.

There is one thing that is a little bit embarrassing for large problems tests so far. FW is fast in each iteration because it requires solving Singular Value Decomposition with only the largest singular value/vector. However, Armadillo doesn't have it now for dense matrices. Stephen used to say he would add it in Armadillo, but I guess he is too busy. So the current FW code performs a full SVD. (This kind of prohibit a large scale problem tests comparison for now.)

In the current unit test file, I tried the example Stephen saved for SDP solver. I manually tuned the iteration numbers for FW and the step size in the projected gradient step (the weird numbers Stephen asked in his comments). Stephen also suggested to try some other examples. I will try it later to see if we can find a better example that is in favor of FW.

Sorry that I am really busy recently, I will definitely fix the problems all you guys mentioned soon. (Hopefully this weekend.) Thanks a lot for the suggestions. 😄

Reply:
No worries about being busy, sometimes life gets in the way. :)  I think maybe we should open a bug noting that `eigs()` can only be used with sparse matrices.  It should be pretty easy to fix actually, then we can just send a patch upstream to Conrad.

If the convergence is slow for the test, I think that is okay---the only goal of the test is to see if it works.  So I think it is okay to make the problem even smaller, or limit the number of iterations to get the test time down, as long as the result conditions we are checking are still meaningful.

For large-scale simulations, if you have a belief that FW will be faster than SDP, I think that is ok with me, and we can confirm that later once dense `eigs()` is available.  (If you are interested in doing that too, I can show you how, just let me know.)  I just want to make sure we are not going to commit code that will be slower than SDP in all cases, otherwise there is no reason for anyone to use it. :)